### ISSUE TITLE
Looking for a bootstrapping option to set the hostname inside a VM.

### ISSUE THREAD

**Issue by mschnell42** [CenturyLinkCloud/chef-provisioning-vsphere#20]
All,
Prior to migrating to chef-provisioning-vsphere, I used knife-vsphere.  In knife-vsphere there seems to be several more VMware options available compared to chef-provisioning-vsphere.  (See https://github.com/ezrapagel/knife-vsphere#knife-vsphere-vm-clone-new-vm-name-template-source-template-name-cspec-customization_spec for that list.)

In particular, I really want to be able to set the hostname of the Ubuntu instance that I spin up in vCenter. At the moment, there does not seem to be a [:customization_spec][:hostname] bootstrapping option. Is there any roadmap for "--chostname CUST_HOSTNAME" (or any other, like "--dest-folder FOLDER" or "--cspec CUST_SPEC" or "--ctz CUST_TIMEZONE") option support?

[With a little bit of guidance, being I'm a relative Chef/chef-provisoining/Ruby newbie, I'd even be willing to help with the any options expansion efforts. ]

Cheers...   Mark

**Comment by mwrock** [comment]
Of the settings you mentioned, only one, `CUST_SPEC` is not implemented. The hostname is set from the `machine` resource name. Also find these settings mentioned in the readme:
```
[:customization_spec][:win_time_zone]
[:vm_folder]
```
Looking over the code, I do see that there once was the ability to pass the name of a vmware customization spec but we never used that here and it looks like that would not work but would be easy to fix.

Let me know if the above options dont work for you or if there are others you would like to see.

Of coarse PRs are welcome and I'd be willing to point you in the right direction if you'd like help getting started.

**Comment by mschnell42** [comment]
Thanks for the feedback.

re: [:vm_folder]...   yep. missed that one.

re: [:customization_spec][:win_time_zone]...  saw it, but ignored it since I'm in a Ubuntu world, not Win.  If this bootstrap options works for Linux also, great.  But, if so, perhaps at least a comment enhancement/expansion on this option telling about Linux support would be appropriate.  Ideally, I'd suggest adding a parallel option that does not have "win_" in the option name.

re: the machine resource name...   well, for me, the machine resource name is getting propagated to...
- the Chef node name: visible in the Chef Manage dashboard and via knife search commands
- the VM name: visible in vCenter
- But it is not propagated into the Ubuntu image itself as a hostname. (I still get the default hostname that my imported template had in it.)

re: CUST_SPEC...   is not something that I specifically need.  It was just a VMware option that was not in chef-provisioning-vsphere.

Any hints/suggestions on how to get the hostname customization work would much appreciated. 

Cheers....  Mark

**Comment by mschnell42** [comment]
So, is the independent setting the hostname in a Linux VM a new feature request or should it work already?

**Comment by mwrock** [comment]
I'm just looking over this code now and noticed a couple things:

We do support `[:customization_spec][:time_zone]` for linux which expects completely different values than windows. I have updated the readme to reflect this.

We also support a `[:customization_spec][:hostname]` which if not present, defaults to the machine resource name. I have also updated the readme to reflect that as well. However, it sounds like thats not working for you. I have found that depending on the version of vsphere and the linux distro you are using, some customization settings do not get set as you would expect.

For example, this driver has some shameful hacks to accommodate ubuntu >= 12.04. Vsphere 5.0/5.5 does not edit the `/etc/interfaces` with the appropriate dns stuff.

If you run your provisioning recipe with debug logging, it will spit out a string representation of all config info it sends to vsphere. You'd want to look for `CustomizationLinuxPrep` and then you would see the properties sent. I'd be curious to know you see the hostname there. If not there is definitely a bug in the driver. If you do (and I suspect you will), let us know the distro and version and perhaps we could add more shameful hacks or refactor  to make streamline adding distro specific tweaks.

**Comment by mschnell42** [comment]
Sorry for the delay...

Hmmm....   It seems like I might be tripping over the "...shameful hacks to accommodate ubuntu >= 12.04. Vsphere 5.0/5.5 does not edit the /etc/interfaces with the appropriate dns stuff." hiccups.   On my client I am running Ubuntu 14.04.02

Short version:
I can't get any of these spec options; in any combination; [:hostname], [:time_zone], nor [:domain]) to work.

Long version:
I used -l debug on my chef-client command. No CustomizationLinuxPrep in the DEBUG output.

The errors that I get are either...
1)   For :hostname, :time_zone, :hostname & :time_zone
```
      ArgumentError
      -------------
      domain is required
```
or 2)  For any combination that includes :domain
```
      NoMethodError
      -------------
      undefined method `[]' for nil:NilClass
```

The Compiled Resources: list (to my lightly trained eye) looks reasonable...
```
[2015-08-12T12:33:23-04:00] DEBUG: ---- End HTTP Status/Header Data ----
[2015-08-12T12:33:23-04:00] DEBUG: ---- HTTP Response Body ----
[2015-08-12T12:33:23-04:00] DEBUG: {"error":["node 'OP-vm-4581' not found"]}
[2015-08-12T12:33:23-04:00] DEBUG: ---- End HTTP Response Body -----
[2015-08-12T12:33:23-04:00] DEBUG: Chef::HTTP calling Chef::HTTP::RemoteRequestID#handle_response
[2015-08-12T12:33:23-04:00] DEBUG: Chef::HTTP calling Chef::HTTP::Authenticator#handle_response
[2015-08-12T12:33:23-04:00] DEBUG: Chef::HTTP calling Chef::HTTP::Decompressor#handle_response
[2015-08-12T12:33:23-04:00] DEBUG: Chef::HTTP calling Chef::HTTP::CookieManager#handle_response
[2015-08-12T12:33:23-04:00] DEBUG: Chef::HTTP calling Chef::HTTP::JSONOutput#handle_response
[2015-08-12T12:33:23-04:00] DEBUG: Expected JSON response, but got content-type ''
[2015-08-12T12:33:23-04:00] DEBUG: Chef::HTTP calling Chef::HTTP::JSONInput#handle_response
[2015-08-12T12:33:23-04:00] INFO: HTTP Request Returned 404 Object Not Found: error

    - creating machine OP-vm-4581 on vsphere://10.65.42.3/sdk?use_ssl=true&insecure=true
    -   datacenter: "SAI Lab"
    -   network_name: "SAI Network 48 - SDN"
    -   resource_pool: "10.65.41.2"
    -   template_name: "GP-small-u14-dhcp---Mark"
    -   num_cpus: "2"
    -   memory_mb: "4096"
    -   customization_spec: {:hostname=>"OP-vm-4581", :time_zone=>"America/Detroit"}
    -   ssh: {:user=>"root", :password=>"Extreme1", :paranoid=>false}establishing connection to 10.65.42.3
[2015-08-12T12:33:23-04:00] DEBUG: Finding pool: 10.65.41.2
[2015-08-12T12:33:23-04:00] DEBUG: Identifying entity part: 10.65.41.2 in folder type: Folder
[2015-08-12T12:33:23-04:00] DEBUG: Parent folder is a folder

    - finding networks...
    - network: SAI Network 48 - SDN is a Network
    - changing template nic for SAI Network 48 - SDN
    ================================================================================
    Error executing action `converge` on resource 'machine[OP-vm-4581]'
    ================================================================================
    
    ArgumentError
    -------------
    domain is required
    
    Resource Declaration:
    ---------------------
    # In /Users/mschnell/.chef/cache/cookbooks/provision/recipes/ig-db.rb
    
     52: machine "OP-vm-#{VMid}" do
     53:   chef_environment 'demo1'
     54:   tag 'cluster.db'
     55:   tag 'db.instance'
     56:   run_list ['ig-db::default']
     57: end
    
    Compiled Resource:
    ------------------
    # Declared in /Users/mschnell/.chef/cache/cookbooks/provision/recipes/ig-db.rb:52:in `from_file'
    
    machine("OP-vm-4581") do
      action :converge
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      chef_environment "demo1"
      chef_server {:chef_server_url=>"https://api.opscode.com/organizations/extreme1", :options=>{:client_name=>"mschnell", :signing_key_filename=>"/Users/mschnell/.chef/mschnell.pem"}}
      driver "vsphere://10.65.42.3/sdk?use_ssl=true&insecure=true"
      machine_options {:bootstrap_options=>{:datacenter=>"SAI Lab", :network_name=>"SAI Network 48 - SDN", :resource_pool=>"10.65.41.2", :template_name=>"GP-small-u14-dhcp---Mark", :num_cpus=>"2", :memory_mb=>"4096", :customization_spec=>{:hostname=>"OP-vm-4581", :time_zone=>"America/Detroit"}, :ssh=>{:user=>"root", :password=>"Extreme1", :paranoid=>false}}}
      declared_type :machine
      cookbook_name "provision"
      recipe_name "ig-db"
      attribute_modifiers [["tags", #<Proc:0x007fe514ed48c8@/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/cheffish-1.2/lib/cheffish.rb:149>], ["tags", #<Proc:0x007fe514ed4698@/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/cheffish-1.2/lib/cheffish.rb:149>]]
      run_list ["ig-db::default"]
    end
    
[2015-08-12T12:33:23-04:00] INFO: Running queued delayed notifications before re-raising exception
[2015-08-12T12:33:23-04:00] DEBUG: Re-raising exception: ArgumentError - machine[OP-vm-4581] (provision::ig-db line 52) had an error: ArgumentError: domain is required
/Users/mschnell/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.7.0/lib/chef/provisioning/vsphere_driver/clone_spec_builder.rb:94:in `customization_options_from'
  /Users/mschnell/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.7.0/lib/chef/provisioning/vsphere_driver/clone_spec_builder.rb:44:in `build'
  /Users/mschnell/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.7.0/lib/chef/provisioning/vsphere_driver/driver.rb:499:in `clone_vm'
  /Users/mschnell/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.7.0/lib/chef/provisioning/vsphere_driver/driver.rb:177:in `find_or_create_vm'
  /Users/mschnell/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.7.0/lib/chef/provisioning/vsphere_driver/driver.rb:135:in `allocate_machine'
  /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-provisioning-1.1.1/lib/chef/provider/machine.rb:30:in `block in <class:Machine>'
  /opt/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `instance_eval'
  /opt/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `recipe_eval_with_update_check'
  /opt/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:45:in `block in action'
  /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-provisioning-1.1.1/lib/chef/provider/machine.rb:37:in `block in <class:Machine>'
  /opt/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `instance_eval'
  /opt/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `recipe_eval_with_update_check'
  /opt/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:45:in `block in action'
  /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-provisioning-1.1.1/lib/chef/provider/machine.rb:55:in `block in <class:Machine>'
  /opt/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `instance_eval'
  /opt/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `recipe_eval_with_update_check'
  /opt/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:45:in `block in action'
  /opt/chefdk/embedded/apps/chef/lib/chef/provider.rb:130:in `run_action'
  /opt/chefdk/embedded/apps/chef/lib/chef/resource.rb:562:in `run_action'
  /opt/chefdk/embedded/apps/chef/lib/chef/runner.rb:49:in `run_action'
  /opt/chefdk/embedded/apps/chef/lib/chef/runner.rb:81:in `block (2 levels) in converge'
  /opt/chefdk/embedded/apps/chef/lib/chef/runner.rb:81:in `each'
  /opt/chefdk/embedded/apps/chef/lib/chef/runner.rb:81:in `block in converge'
  /opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/resource_list.rb:83:in `block in execute_each_resource'
  /opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:116:in `call'
  /opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:116:in `call_iterator_block'
  /opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
  /opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:104:in `iterate'
  /opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
  /opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/resource_list.rb:81:in `execute_each_resource'
  /opt/chefdk/embedded/apps/chef/lib/chef/runner.rb:80:in `converge'
  /opt/chefdk/embedded/apps/chef/lib/chef/client.rb:339:in `block in converge'
  /opt/chefdk/embedded/apps/chef/lib/chef/client.rb:334:in `catch'
  /opt/chefdk/embedded/apps/chef/lib/chef/client.rb:334:in `converge'
  /opt/chefdk/embedded/apps/chef/lib/chef/client.rb:353:in `converge_and_save'
  /opt/chefdk/embedded/apps/chef/lib/chef/client.rb:457:in `run'
  /opt/chefdk/embedded/apps/chef/lib/chef/application.rb:271:in `block in fork_chef_client'
  /opt/chefdk/embedded/apps/chef/lib/chef/application.rb:259:in `fork'
  /opt/chefdk/embedded/apps/chef/lib/chef/application.rb:259:in `fork_chef_client'
  /opt/chefdk/embedded/apps/chef/lib/chef/application.rb:225:in `block in run_chef_client'
  /opt/chefdk/embedded/apps/chef/lib/chef/local_mode.rb:39:in `with_server_connectivity'
  /opt/chefdk/embedded/apps/chef/lib/chef/application.rb:213:in `run_chef_client'
  /opt/chefdk/embedded/apps/chef/lib/chef/application/client.rb:402:in `block in interval_run_chef_client'
  /opt/chefdk/embedded/apps/chef/lib/chef/application/client.rb:392:in `loop'
  /opt/chefdk/embedded/apps/chef/lib/chef/application/client.rb:392:in `interval_run_chef_client'
  /opt/chefdk/embedded/apps/chef/lib/chef/application/client.rb:382:in `run_application'
  /opt/chefdk/embedded/apps/chef/lib/chef/application.rb:60:in `run'
  /opt/chefdk/embedded/apps/chef/bin/chef-client:26:in `<top (required)>'
  /usr/bin/chef-client:51:in `load'
  /usr/bin/chef-client:51:in `<main>'

Running handlers:
[2015-08-12T12:33:23-04:00] ERROR: Running exception handlers
Running handlers complete
[2015-08-12T12:33:23-04:00] ERROR: Exception handlers complete
```

Bottom line:
I'll start poking around the driver code.  But, being a newbie to Chef, Ruby & chef-provisioning-vsphere, any pointer to the "right area" to look at would be welcome.  ;-)

**Comment by mwrock** [comment]
My apologies for the delay. The `domain is required` error is expected. The underlying vsphere API `CustomizationLinuxPrep` requires it. However I am concerned with the `undefined method `[]' for nil:NilClass` error and have not been able to reproduce locally.

I'd really like to get the full stacktrace for that error. If you have a chance, could you add the domain and run this in debug again and provide the output?

**Comment by mwrock** [comment]
Never mind. I have a repro now.
